src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,1026 @@
+                f" {self.num_heads})."
+            )
+
+        # Layer-wise attention scaling
+        self.layer_number = max(1, layer_number)
+        coeff = self.layer_number
+        self.norm_factor = math.sqrt(self.head_dim) * coeff
+
+        # Scaled Softmax
+        self.scale_mask_softmax = ScaledSoftmax(
+            self.fp16,
+            self.bf16,
+            self.masked_softmax_fusion,
+            attention_mask_func,
+            self.attention_softmax_in_fp32,
+            coeff,
+            # To avoid cache problems, we use a small offset when creating the alibi positional embeddings
+            config.seq_length + config.offset_alibi,
+        )
+
+        self.query_key_value = nn.Linear(self.hidden_size, 3 * self.hidden_size, dtype=dtype, bias=True)
+        self.dense = nn.Linear(self.hidden_size, self.hidden_size, dtype=dtype)
+        self.skip_bias_add = config.skip_bias_add
+        self.skip_bias_add_qkv = config.skip_bias_add_qkv
+        self.attention_dropout = torch.nn.Dropout(config.attention_dropout)
+
+    def forward(
+        self,
+        hidden_states,
+        layer_past=None,
+        attention_mask=None,
+        alibi=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+        # hidden_states: [batch_size, seq_length, hidden_size]
+        # repeat alibi tensor with the batch size
+        alibi = alibi.repeat(hidden_states.shape[0], 1, 1).to(hidden_states.device)
+
+        bias = self.query_key_value.bias if not self.skip_bias_add_qkv else None
+
+        output_bias = self.query_key_value.bias if self.skip_bias_add_qkv else None
+
+        mixed_x_layer = F.linear(hidden_states, self.query_key_value.weight, bias)
+
+        # [batch_size, seq_length, 3 x hidden_size] --> [batch_size, seq_length, num_heads, 3 x head_dim]
+        new_tensor_shape = mixed_x_layer.size()[:-1] + (self.num_heads, 3 * self.head_dim)
+        mixed_x_layer = mixed_x_layer.view(*new_tensor_shape)
+
+        # [batch_size, seq_length, num_heads, 3 x head_dim] --> 3  [batch_size, seq_length, num_heads, head_dim]
+        (query_layer, key_layer, value_layer) = split_tensor_along_last_dim(mixed_x_layer, 3)
+
+        if layer_past is not None:
+            past_key, past_value = layer_past
+            key_layer = torch.cat((past_key.type_as(key_layer), key_layer), dim=1)
+            value_layer = torch.cat((past_value.type_as(value_layer), value_layer), dim=1)
+
+        if use_cache is True:
+            present = (key_layer, value_layer)
+        else:
+            present = None
+
+        # [batch_size, head_dim, q_length, k_length]
+        output_size = (query_layer.size(0), query_layer.size(2), query_layer.size(1), key_layer.size(1))
+
+        # [batch_size, q_length, num_heads, head_dim] -> [q_length, batch_size * num_heads, head_dim]
+        # query_layer = query_layer.contiguous().view(output_size[2], output_size[0] * output_size[1], -1)
```

Comment:
Are those comments necessary? I understand the first one, but what about the second one?

Reply:
There are a lot of comments here and below, which do not seem to help comprehension. Do we need them?